pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,241 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	registerapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha1"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// RegisterCallbackFn is the type of the callback function that handlers will provide
+type RegisterCallbackFn func(pluginName string, versions []string, socketPath string) error
+
+// Watcher is the plugin watcher
+type Watcher struct {
+	path     string
+	handlers map[string]RegisterCallbackFn
+	stopCh   chan interface{}
+	fs       utilfs.Filesystem
+	watcher  *fsnotify.Watcher
+	wg       sync.WaitGroup
+	mutex    sync.Mutex
+}
+
+// NewWatcher provides a new watcher
+func NewWatcher(sockDir string) Watcher {
+	return Watcher{
+		path:     sockDir,
+		handlers: make(map[string]RegisterCallbackFn),
+		fs:       &utilfs.DefaultFs{},
+	}
+}
+
+// AddHandler registers a callback to be invoked for a particular type of plugin
+func (w *Watcher) AddHandler(handlerType string, handlerCbkFn RegisterCallbackFn) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	w.handlers[handlerType] = handlerCbkFn
+}
+
+// Creates the plugin directory, if it doesn't already exist.
+func (w *Watcher) createPluginDir() error {
+	glog.Infof("Ensuring Plugin directory at %s ", w.path)
+	err := w.fs.MkdirAll(w.path, 0755)
+	if err != nil {
+		return fmt.Errorf("error (re-)creating driver directory: %s", err)
+	}
+	return nil
+}
+
+// Walks through the plugin directory to discover any existing plugin sockets.
+func (w *Watcher) traversePluginDir() error {
+	files, err := w.fs.ReadDir(w.path)
+	if err != nil {
+		return fmt.Errorf("error reading the plugin directory: %v", err)
+	}
+	for _, f := range files {
+		// Currently only supports flat fs namespace under the plugin directory.
+		// TODO: adds support for hierarchical fs namespace.
+		if !f.IsDir() && filepath.Base(f.Name())[0] != '.' {
+			w.registerPlugin(path.Join(w.path, f.Name()))
+		}
+	}
+	return nil
+}
+
+func (w *Watcher) init() error {
+	if err := w.createPluginDir(); err != nil {
+		return err
+	}
+	return w.traversePluginDir()
+}
+
+func (w *Watcher) registerPlugin(socketPath string) error {
+	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
+	glog.V(2).Infof("registerPlugin called for socketPath: %s", socketPath)
+	client, conn, err := dial(socketPath)
+	if err != nil {
+		return fmt.Errorf("dial failed at socket %s, err: %v", socketPath, err)
+	}
+	defer conn.Close()
+	infoResp, err := client.GetInfo(context.Background(), &registerapi.InfoRequest{})
```

Comment:
One last thing can we add a timeout on the context?
A one second timeout seems reasonable :)